android/app/src/main/java/host/exp/exponent/MainActivity.java

```diff
@@ -25,20 +44,36 @@ public String developmentUrl() {
     return DetachBuildConstants.DEVELOPMENT_URL;
   }
 
-  @Override
-  public List<String> sdkVersions() {
-    return new ArrayList<>(Arrays.asList("31.0.0"));
-  }
-
   @Override
   public List<ReactPackage> reactPackages() {
     return ((MainApplication) getApplication()).getPackages();
   }
 
   @Override
   public List<Package> expoPackages() {
-    // Here you can add your own packages.
-    return super.expoPackages();
```

Comment:
if people are upgrading from an older SDK version and leave this in their MainActivity, will it have the expected behavior (this will be null and then ExperiencePackagePicker will be used)? Want to minimize the upgrade pain if possible.

Reply:
`MainActivity` extends `DetachActivity` in which I made `expoPackages` an abstract method so people have to define their own packages. Why?

Both `DetachActivity` and `ExponentPackagePicker` live in `expoview`, which only _softly_ depends on all those universal modules — if a developer removed `api 'host.exp.exponent:unimodule'` from `app/build.gradle`, application would crash in runtime with `ClassNotFoundException`, because `ExponentPackagePicker` would try to instantiate module's package.

One solution for this crash would be to replace `import class;` in `ExponentPackagePicker` with reflection and `Class.forName`, but I'm not sure it will be a real win in the long run (I don't know if we would be able to return `List<Package>` in that case).

Another solution would be to let ProGuard manage which classes end up in the final package, but that would make debug and release builds differ more, not really in a predictable way.

Yet another solution would be to rename current `DetachActivity` to `BaseDetachActivity`, make `MainActivity` extend `BaseDetachActivity` by default and ask the users to “just download this `DetachActivity` to their projects”, but I'm not sure this would be anyhow better than copy-pasting this fragment.

----

I totally agree relieving upgrade pain for the developers is extremely important for people to happily upgrade their projects (and that, in fact, is crucial for us for many reasons), but I also sense that adding explicit list of the packages to `MainActivity` may be a good PR for Expo!

I think one of the reasons people may turn down developing with Expo is some amount of uncustomizable magic happening behind the scenes. On one hand this is one of the key features of Expo, on the other hand I've seen people arguing against Expo where I believe this was their pain point. Adding an explicit list of `expoPackages`, for these people, may be a sign that power over their project is slowly moving into their hands.

However, some people may not appreciate this and would rather just use `super.expoPackages()`… What do you think? 😃

Reply:
Thanks for the detailed explanation. I agree that making it very explicit that people can control these things now could be a good thing, and I think your assessment of the tradeoffs of the other options makes sense. I'm thinking that sticking with this solution is probably the best way for now, so let's go with that. 👍